File: backend/app/schemas/cultivation.py
"""修习之路 Pydantic Schemas"""

from pydantic import BaseModel, Field
from typing import Optional
from datetime import date, datetime


class SkillTreeProgress(BaseModel):
    """单条技能树进度"""
    tree_name: str           # "鉴宝"
    label: str               # "鉴宝之路"
    icon: str                # emoji
    level: int               # 1-5
    current: int             # 当前进度数值
    threshold: int           # 升级所需阈值
    percentage: float        # 0-100


class CultivationStatus(BaseModel):
    """用户修习状态"""
    xp: int
    rank: str                # 初窥门径 / 略有小成 / 融会贯通 / 炉火纯青 / 一代宗师
    rank_index: int          # 0-4
    xp_to_next: int          # 距离下一段位所需 XP
    skill_trees: list[SkillTreeProgress]  # 6 条


class DailyQuest(BaseModel):
    """每日任务"""
    id: int
    quest_template_id: str
    title: str
    description: str
    module: str
    skill_tree: str
    xp_reward: int
    status: str              # pending | completed | claimed
    icon: str = "📋"


class WeeklyChallenge(BaseModel):
    """每周挑战"""
    week_label: str          # "2026-W27"
    theme: str               # "青瓷周"
    description: str
    tasks_completed: int
    tasks_total: int
    reward_stamp_name: str   # 限定印章名
    reward_stamp_icon: str   # 印章图标
    expires_at: str          # ISO 日期


class QuestCompleteResponse(BaseModel):
    """任务完成响应"""
    xp_gained: int
    total_xp: int
    new_rank: Optional[str] = None      # 升级后的段位，未升级则为 None
    new_rank_index: Optional[int] = None
    stamp_earned: Optional[dict] = None  # 获得的新印章信息

File: frontend/src/services/cultivation.ts
/** 非遗修习之路 API */

import api from './api'

export interface SkillTreeProgress {
  tree_name: string
  label: string
  icon: string
  level: number
  current: number
  threshold: number
  percentage: number
}

export interface CultivationStatus {
  xp: number
  rank: string
  rank_index: number
  xp_to_next: number
  skill_trees: SkillTreeProgress[]
  streak_days: number
  longest_streak: number
  last_active_date: string | null
  streak_bonus_active: boolean
}

export interface DailyQuest {
  id: number
  quest_template_id: string
  title: string
  description: string
  module: string
  skill_tree: string
  xp_reward: number
  status: 'pending' | 'completed' | 'claimed'
  icon: string
  condition_type: string
  condition_threshold: number
  condition_progress: number
}

export interface WeeklyChallenge {
  week_label: string
  theme: string
  description: string
  tasks_completed: number
  tasks_total: number
  reward_stamp_name: string
  reward_stamp_icon: string
  expires_at: string
}

export interface QuestCompleteResponse {
  xp_gained: number
  total_xp: number
  new_rank: string | null
  new_rank_index: number | null
  stamp_earned: Record<string, unknown> | null
}

export interface StreakInfo {
  streak_days: number
  longest_streak: number
  last_active_date: string | null
  streak_bonus_active: boolean
}

export interface QuestCompletedItem {
  quest_id: number
  title: string
  xp_gained: number
  skill_tree: string
  icon: string
}

export interface QuestProgressUpdate {
  id: number
  condition_progress: number
  condition_threshold: number
  status: string
}

export interface QuestCheckResponse {
  quests_completed: QuestCompletedItem[]
  total_xp_gained: number
  new_rank: string | null
  quests_updated: QuestProgressUpdate[]
}

/** 获取修习状态 */
export async function getCultivationStatus(): Promise<CultivationStatus> {
  const { data } = await api.get('/api/cultivation/status')
  return data
}

/** 获取今日任务 */
export async function getDailyQuests(): Promise<DailyQuest[]> {
  const { data } = await api.get('/api/cultivation/quests')
  return data
}

/** 手动完成任务（仅不可追踪任务） */
export async function completeQuest(questId: number): Promise<QuestCompleteResponse> {
  const { data } = await api.post(`/api/cultivation/quests/${questId}/complete`)
  return data
}

/** 触发任务自动结算 */
export async function checkQuests(): Promise<QuestCheckResponse> {
  const { data } = await api.post('/api/cultivation/quests/check')
  return data
}

/** 获取每周挑战 */
export async function getWeeklyChallenge(): Promise<WeeklyChallenge> {
  const { data } = await api.get('/api/cultivation/weekly-challenge')
  return data
}

/** 获取连胜状态 */
export async function getStreak(): Promise<StreakInfo> {
  const { data } = await api.get('/api/cultivation/streak')
  return data
}
